### ISSUE TITLE
Upgrade SQLitePclRaw to resolve the Android N sqlite incompatibility

### ISSUE THREAD

**Issue by ghuntley** [akavache/Akavache#306]

**Comment by ajuv001** [comment]
I am using Akavache ver 4.1.2 in my app. I have updated Android SDK and Xamarin to support Android N. After that I am getting an exception saying DllNotFoundException. I am getting exception on the following statement "BlobCache.LocalMachine". 
"builder.RegisterInstance(BlobCache.LocalMachine).As<IBlobCache>().ExternallyOwned();".

**Comment by rleffler** [comment]
I am also having this same issue.  Is there any workaround for this?

**Comment by codercampos** [comment]
There is a pull request where the reference has been updated. We have to wait until akavache 5.0 ships out or build that commit by yourself (if is it possible)

**Comment by rleffler** [comment]
Thanks, any idea when akavache 5.0 might be released?

**Comment by codercampos** [comment]
Not quite sure. I'm also waiting for this release because I want to update to Android N on Xamarin and I can't until akavache updates its nuget package

**Comment by flagbug** [comment]
@codercampos @rleffler Why can't you just upgrade the SQLite dependency?

**Comment by codercampos** [comment]
From Nuget says it depends on SQLiteRaw_basic which has been deprecated. I will try to force it tomorrow (at least on Xamarin, that's the problem)

**Comment by codercampos** [comment]
@flagbug wrong name, SQLitePCL.raw_basic is the package (that is supposed to be the problem) I will try add a plugin that the author suggest

**Comment by kjeremy** [comment]
I'm hitting this now with the new images from Google.

**Comment by ericsink** [comment]
Akavache needs to be transitioned to SQLitePCL.raw 1.0, which was released today.

The package ids for SQLitePCL.raw have changed.  And there is an initialization step required.

You should probably be able to just change your dependency from SQLitePCL.raw_basic to SQLitePCLRaw.bundle_e_sqlite3 (or if you want to use the system-provided SQLite on iOS, use bundle_green).

And add a call to:
 
    SQLitePCL.Batteries.Init();

Lots more details in the README for SQLitePCL.raw.

Do let me know if you have I can help with the transition.

**Comment by maxchen666** [comment]
Hi Everyone, do you have any update on this issue? Currently, on Android 7, it pops up an annoying message.

**Comment by ericsink** [comment]
FWIW, more docs on the SQLitePCL.Batteries.Init() issue:

https://github.com/ericsink/SQLitePCL.raw/wiki/SQLitePCL.Batteries.Init

**Comment by PureWeen** [comment]
Yea I was pretty excited about the bundles :-)  

Is it necessary to recompile Akavache against the new libraries? Like all I've done in my projects is force remove the "SQLitePCL.raw_basic" dependency and then add in 
https://www.nuget.org/packages/SQLitePCL.raw/0.9.3
And it all seems to work fine

And I was just going to do the same thing with the green bundle... Or would you recommend recompiling the Akavache sources? For example would it be recommended to recompile Akavache.SQLite3 (PCL) against the new 1.0 PCL or does it work fine to just install the bundles?

**Comment by ericsink** [comment]
It had no occurred to me before, but now that you mention it, yes, I suppose there are certain situations where if you force remove the old pkg and the new one, it might work.

But you can think of that as a happy fluke.

And I am disinclined to spend time thinking about which scenarios happen to work and why.  The best approach is to recompile.

For the do-it-yourself-er, I do think that the following should work:

- Rebuild Akavache.SQLite3 against SQLitePCLRaw.core 1.0
- In your app project, add one of the bundles and call SQLitePCL.Batteries.Init()

In other words, in terms of the actual SQLite APIs, nothing broke between 0.7 and 1.0.  All the changes were in package ids, package layout, and initialization.

**Comment by PureWeen** [comment]
Hehe I had a feeling that would be the response :-) As of now our apps been working fine so yay for happy flukes... 

I haven't fully tested this yet but I do have a fork of Akavache 
https://github.com/PureWeen/Akavache

Where I basically have all the nuget packages updated and I installed the GREEN bundles. Only because using the provided iOS libraries sounded good to me (which probably isn't a great reason)

I added the battery inits into the Registrations.cs files because that seemed like a great place for it... 
https://github.com/PureWeen/Akavache/blob/master/Akavache/Registrations.cs

Thus making it fall into this category
"Maybe your library wants to use SQLite as merely an implementation detail that is not visible to users."

Here's a link to the compiled dlls
https://github.com/PureWeen/Akavache/tree/master/Library

As of now I've only really tested any of this on monoandroid and so far it seems to work fine...
I'll test it on iOS soon here and probably won't do any meaningful test against Windows for some time :-/

**Comment by dhaligas** [comment]
@ghuntley what is the status of this?

**Comment by PureWeen** [comment]
I made this one branched off of the akavache5-master branch

https://github.com/PureWeen/Akavache/tree/akavache5-master-sqllite-unopinionated

Mainly because as the sqllite versions kept incrementing I figured it'd just be simpler to  have a version of Akavache that only has the PCL library in it and then I'll just worry about the bundles myself. 

You can either Init the batteries yourself or pass in an action 

```
 BlobCache.Init("TestRunner",
                   () =>
                   {
                       SQLitePCL.Batteries_V2.Init();
                   });
```

I've used it on iOS and Android and it seems to work fine so far :-)

**Comment by dhaligas** [comment]
@PureWeen do you have a nuget for it? or can you point me to the dlls and nuget packages you are using?

**Comment by PureWeen** [comment]
https://github.com/PureWeen/Akavache/tree/akavache5-master-sqllite-unopinionated/Akavache/Libraries

or this one just has the green 1.0.1 version of the batteries compiled in
https://github.com/PureWeen/Akavache/tree/master/Library

I haven't done a lot of testing with these though :-) so use at your own risk

Here's a good doc about how all the batteries things work
https://github.com/ericsink/SQLitePCL.raw/wiki/SQLitePCL.Batteries.Init

**Comment by dhaligas** [comment]
@PureWeen thanks.  I see those have been updated to 1.1.0.  any plans to update your branch?

**Comment by PureWeen** [comment]
Not currently.

I created
https://github.com/PureWeen/Akavache/tree/akavache5-master-sqllite-unopinionated

Specifically because I didn't want to keep the other one up to date :-)

**Comment by dhaligas** [comment]
@PureWeen thanks this worked for me

**Comment by ghuntley** [comment]
Resolved at https://github.com/akavache/Akavache/pull/291#issuecomment-253601110 please test using the packages on myget.

**Comment by logikonline** [comment]
I am not seeing the ability to reference SQLiteAsyncConnection.  Was this removed from Akavache 5?

**Comment by logikonline** [comment]
Thank you for updating the library but we previously accessed the SQLiteAsyncConnection from Akavache.Sqlite3.Internal which we see is no longer accessible.  This was vital for us to provide the seamless use of the library.

Reading the documentation from Android - is there some steps we can do to allow us to use this object?  Can Akavache expose this or can you provide us with the steps necessary to continue using this?

Any assistance is greatly appreciated.

**Comment by ghuntley** [comment]
Open a new discussion please.
